File: php/raindrops/raindrops.php
<?php

function raindrops($n) {
  $rules = [
    ["3", "Pling"],
    ["5", "Plang"],
    ["7", "Plong"],
  ];

  $makeSound = function($rule) use ($n) {
    list($factor, $sound) = $rule;
    return $n % $factor == 0 ? $sound : "";
  };

  $sounds = implode(array_map($makeSound, $rules));
  return empty($sounds) ? "$n" : $sounds;
}

File: kotlin/raindrops/src/main/kotlin/Raindrops.kt
typealias Rule = Pair<Int, String>

val RULES: List<Rule> = listOf(Pair(3, "Pling"),
                               Pair(5, "Plang"),
                               Pair(7, "Plong"))

private fun ruleSound(n: Int, rule: Rule): String {
  val (factor, sound) = rule
  return if (n % factor == 0) { sound } else { "" }
}

object Raindrops {
  fun convert(n: Int): String {
    return RULES.fold("") { sounds, rule -> sounds + ruleSound(n, rule) }.let {
      if (it.isBlank()) { n.toString() } else { it }
    }
  }
}
